### DISCUSSION TITLE
Dummy Regressor returning the value in t-1?

### DISCUSSION THREAD

**Discussion by XaviMartiCamarasa** [online-ml/river#1090] [Q&A] [answered]
Hi.
Is there any object/regressor that returns the value  in t-1?
I'm trying to build a Dummy Regressor

Many Thanks

**Comment by MaxHalford** [comment, marked as answer]
Hey there. This should work:

```py
from river import dummy, stats

model = dummy.StatisticRegressor(stats.Shift(1))
```

**Reply by XaviMartiCamarasa** [reply]
Just doing it! Many thanks Max!

**Reply by XaviMartiCamarasa** [reply]
Hi Max.
I'm getting this error when doing that. Thats the first value
*********************************************************************
y: 168
y_pred: None
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_14212/3649714652.py in <module>
      6         df_data= data_orig.iloc[:, i]
      7         label=data_orig.columns[i]
----> 8         evaluate_model_without_outliers(model_lin_reg.clone(),df_data,label,ax,"linear")
      9         i += 1
     10 plt.tight_layout()

~\AppData\Local\Temp/ipykernel_14212/1814005244.py in evaluate_model_without_outliers(model, df_data, column, ax, etiqueta)
     81         model.learn_one(x, y)
     82         # Update the error metric
---> 83         metric.update(y, y_pred)
     84 
     85         #get metric value for plotting

~\Anaconda3\lib\site-packages\river\utils\rolling.py in update(self, *args, **kwargs)
     79         if len(self.window) == self.window_size:
     80             self.obj.revert(*self.window[0][0], **self.window[0][1])
---> 81         self.obj.update(*args, **kwargs)
     82         self.window.append((args, kwargs))
     83         return self

~\Anaconda3\lib\site-packages\river\metrics\base.py in update(self, y_true, y_pred, sample_weight)
    335 
    336     def update(self, y_true, y_pred, sample_weight=1.0):
--> 337         self._mean.update(x=self._eval(y_true, y_pred), w=sample_weight)
    338         return self
    339 

~\Anaconda3\lib\site-packages\river\metrics\mae.py in _eval(self, y_true, y_pred)
     30 
     31     def _eval(self, y_true, y_pred):
---> 32         return abs(y_true - y_pred)

TypeError: unsupported operand type(s) for -: 'int' and 'NoneType'
*************************************************************
This  is my code here: 
```

def get_ordinal_date(x):
    print(x.toordinal())
    return {'ordinal_date': x.toordinal()}
    
model_statistic_regressor = compose.Pipeline(
    ('ordinal_date', compose.FuncTransformer(get_ordinal_date)),
    ('scale', preprocessing.StandardScaler()),
    ('statistic_reg', dummy.StatisticRegressor (stats.Shift(1))) 
)


def evaluate_model_without_outliers(model,df_data,column,ax,etiqueta): 

    metric = utils.Rolling(metrics.MAE(), 52)#52 semanas tiene el año que es un ciclo completo

    dates = []
    y_trues = []
    y_preds = []
    sdt=df_data.std()
    mean=df_data.mean()
    outlier_value= mean+(3*sdt)
    num_outliers=0
    i=0
    
    for x, y in df_data.T.iteritems():
     
        # Obtain the prior prediction and update the model in one go
      y_pred = model.predict_one(x)
      print ("x:",x)
      print ("y_pred:",y_pred)
      print ("y:",y)
      print ("y_pred:",y_pred)
      if  (y<=outlier_value) :# Get rid  off outliers
        
        model.learn_one(x, y)
        # Update the error metric
        metric.update(y, y_pred)

        #get metric value for plotting
        if etiqueta=="linear":
          df_error_mae_linear.at[i,column] =metric.get()
        elif etiqueta=='stg':
          df_error_mae_stg.at[i,column] =metric.get()
        elif etiqueta=='hft':
          df_error_mae_hft.at[i,column] =metric.get()
        elif etiqueta=='static':
          df_error_mae_baseline.at[i,column] =metric.get()

        i=i+1
        
        # Store the true value and the prediction
        dates.append(x)
        y_trues.append(y)
        y_preds.append(y_pred)
      else:
        num_outliers=num_outliers+1
       
    
      # Plot the results
  # fig, ax = plt.subplots(figsize=(9, 5))
    value=metric.get()
    
    ax.grid(alpha=0.75)
    ax.plot(dates, y_trues, lw=3, color='#2ecc71', alpha=0.8, label='Ground truth')
    ax.plot(dates, y_preds, lw=3, color='#e74c3c', alpha=0.8, label='Prediction')
    ax.legend()
    ax.set_title(metric) 
    ax.set_xlabel(column)

    #Añadimos filas con los datos para cada columna
    df_datos.at['Num Recs', column] = len(df_data)
    df_datos.at['MAE', column] = value
    df_datos.at['Outliers', column] = num_outliers

```
It looks as if the dummy regressor does not predict anything leading to an error

![image](https://user-images.githubusercontent.com/117338671/201716984-458d449c-d2f2-4c64-85b9-2bc070e07283.png)

What is wrong here?

**Reply by MaxHalford** [reply]
I believe it's because the model always returns `None` at first. You could fix this by overriding the model's `predict_one` method:

```py
from river import dummy

class StatisticRegressor(dummy.StatisticRegressor):

    def predict_one(self, x):
        yp = super().predict_one(x)
        if yp is None:
            return 0
        return yp
```

**Reply by XaviMartiCamarasa** [reply]
Still remains the same:
![image](https://user-images.githubusercontent.com/117338671/201720331-18f7bafc-fc6f-4627-be51-2b0f190d633a.png)

Something I'm doing wrong

**Reply by MaxHalford** [reply]
You're using `dummy.StatisticRegressor` and not `StatisticRegressor` in your pipeline

**Reply by XaviMartiCamarasa** [reply]
Sorry for the mistake Max! :(. It works now perfectly!
